sequencer/src/enriched_handle.rs

```diff
@@ -0,0 +1,55 @@
+use hotshot::{traits::NodeImplementation, types::SystemContextHandle};
+use hotshot_types::{light_client::StateSignature, traits::node_implementation::NodeType};
+use std::{
+    ops::{Deref, DerefMut},
+    sync::{Arc, RwLock},
+};
+
+use crate::api::signature_pool::StateSignaturePool;
+
+pub struct EnrichedSystemContextHandle<TYPES: NodeType, I: NodeImplementation<TYPES>> {
+    handle: SystemContextHandle<TYPES, I>,
+
+    state_signature_pool: Arc<RwLock<StateSignaturePool>>,
+}
+
+impl<TYPES: NodeType, I: NodeImplementation<TYPES>> Deref
+    for EnrichedSystemContextHandle<TYPES, I>
+{
+    type Target = SystemContextHandle<TYPES, I>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.handle
+    }
+}
+
+impl<TYPES: NodeType, I: NodeImplementation<TYPES>> DerefMut
+    for EnrichedSystemContextHandle<TYPES, I>
+{
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.handle
+    }
+}
+
+impl<TYPES: NodeType, I: NodeImplementation<TYPES>> From<SystemContextHandle<TYPES, I>>
+    for EnrichedSystemContextHandle<TYPES, I>
+{
+    fn from(handle: SystemContextHandle<TYPES, I>) -> Self {
+        Self {
+            handle,
+            state_signature_pool: Default::default(),
+        }
+    }
+}
+
+impl<TYPES: NodeType, I: NodeImplementation<TYPES>> EnrichedSystemContextHandle<TYPES, I> {
+    pub fn get_state_signature(&self, height: u64) -> Option<StateSignature> {
+        let pool_guard = self.state_signature_pool.read().unwrap();
+        pool_guard.get_signature(height)
+    }
+
+    pub fn add_new_state_signature(&mut self, height: u64, signature: StateSignature) {
```

Comment:
Not sure what the correct thing should be, but there's no point in using an `RwLock` if we only ever obtain write access through a `&mut`. Seems like this should be `&self` (if this is actually going to be shared between threads) or we should get rid of the `RwLock` (if it's not shared)

Reply:
I'm not familiar with the tide-disco apps. For safety, I kept the `RwLock` and remove those `mut`s. [f7033f7](https://github.com/EspressoSystems/espresso-sequencer/pull/879/commits/f7033f7222880be8e53eeee798c12ca2691f496c)